### PR TITLE
Add some laws

### PR DESCRIPTION
**Pull request by edmundnoble**
A few more laws are incoming, as well as actual tests using those laws.

I included a couple of explanations for the laws, will add a lot more soon.

I split laws between type class laws, homomorphism laws and action laws. I will also need default laws as their own set.

I like the idea of keeping all of our type class laws in one file; they're all very short, and there's a *LOT* of knowledge to be gained by comparing laws. For example, it's easy to understand all of the laws of various kinds of functors when you see them all next to each other.

The laws are very generically specified; they don't require the testz dependency to be written out.

### CODE REVIEW

File: laws/src/main/scala/scalaz/tests/TypeclassLaws.scala [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+object ApplicativeLaws {
+  @inline
+  def applyIdentity[F[_], A, B, T](in: A)(f: A => B)(assert: (F[B], F[B]) => T)(implicit F: Applicative[F]) =
+    assert(
+      F.pure(f(in)),
+      F.ap(F.pure(in))(F.pure(f))
+    )
+}
+
+object EqLaws {
```
Comment (alexknvl) [inline laws/src/main/scala/scalaz/tests/TypeclassLaws.scala]: Reflexivity? Transitivity and symmetry follow from the indiscernability + reflexivity.

Comment (edmundnoble) [reply laws/src/main/scala/scalaz/tests/TypeclassLaws.scala]: Will do.

Comment (edmundnoble) [reply laws/src/main/scala/scalaz/tests/TypeclassLaws.scala]: Done.

File: laws/shared/src/main/scala/scalaz/laws/TypeclassLaws.scala [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+  // Nested `first` applications introducing multiple arguments
+  // can be replaced by a single `first` application introducing a tuple-full of arguments.
+  //
+  // first' . first' ≡ dimap assoc unassoc . first' where
+  // assoc ((a,b),c) = (a,(b,c))
+  // unassoc (a,(b,c)) = ((a,b),c)
+  @inline
+  def firstAssoc[P[_, _], A, B, C, T](
+    in: P[A, A]
+  )(assert: (P[((A, B), C), ((A, B), C)], P[((A, B), C), ((A, B), C)]) => T)(implicit P: Strong[P]): T =
```
[the remark this answers is not in the document]
Comment (edmundnoble) [reply laws/shared/src/main/scala/scalaz/laws/TypeclassLaws.scala]: Type aliases are the devil ;)

File: laws/shared/src/main/scala/scalaz/laws/ActionLaws.scala [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (edmundnoble) [reply laws/shared/src/main/scala/scalaz/laws/ActionLaws.scala]: I removed it in the second PR, I can remove it in this one if you'd prefer.

File: laws/shared/src/main/scala/scalaz/laws/TypeclassLaws.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdegoes) [reply laws/shared/src/main/scala/scalaz/laws/TypeclassLaws.scala]: OTOH `type Law[A] = forall T. ((A, A) => T) => T`

**Pull request merged**